### PR TITLE
investor withdraw

### PR DESCRIPTION
**Pull request by diepmv**

### CODE REVIEW

File: contracts/TokenVesting.sol [hunk 1]
```diff
@@ -266,6 +321,42 @@ contract TokenVesting is Ownable, ReentrancyGuard{
...
+    * @return the total vested amount of tokens of sender
+    */
+    function investorWithdraw(bool keepERC)
+        public
+        nonReentrant returns (uint){
+
+        uint256 count = holdersVestingCount[msg.sender];
+        uint256 total = 0;
+        for (uint i = 0; i < count; i++) {
+            VestingSchedule memory vestingSchedule = getVestingSchedule(computeVestingScheduleIdForAddressAndIndex(msg.sender, i));
```
Comment (Zergity) [inline contracts/TokenVesting.sol:336]: please call the `release` function here, don't rewrite the logic

**Review by Zergity — CHANGES_REQUESTED** [review]

File: contracts/TokenVesting.sol [hunk 2]
```diff
@@ -9,6 +9,61 @@ import "@openzeppelin/contracts/access/Ownable.sol";
...
+    ) internal {
+        // bytes4(keccak256(bytes('transfer(address,uint256)')));
+        (bool success, bytes memory data) = token.call(abi.encodeWithSelector(0xa9059cbb, to, value));
+        require(
+            success && (data.length == 0 || abi.decode(data, (bool))),
+            'TransferHelper::safeTransfer: transfer failed'
+        );
+    }
+
+    function safeTransferFrom(
```
[the remark this answers is not in the document]
Comment (diepmv) [reply contracts/TokenVesting.sol]: done

File: contracts/TokenVesting.sol [hunk 3]
```diff
@@ -266,6 +321,42 @@ contract TokenVesting is Ownable, ReentrancyGuard{
         _token.safeTransfer(beneficiaryPayable, amount);
     }
 
+
+    /**
+    * @notice Withdraw vested amount of tokens from all schedules of sender
+    * @return the total vested amount of tokens of sender
+    */
+    function investorWithdraw(bool keepERC)
```
[the remark this answers is not in the document]
Comment (diepmv) [reply contracts/TokenVesting.sol]: done

File: contracts/TokenVesting.sol (hunk 1 above)
Comment (diepmv) [reply contracts/TokenVesting.sol:295]: I've added some checks

File: contracts/TokenVesting.sol [hunk 4]
```diff
@@ -266,6 +280,46 @@ contract TokenVesting is Ownable, ReentrancyGuard{
...
     }
 
+
+    /**
+    * @notice Withdraw vested amount of tokens from all schedules of sender
+    * @return the total vested amount of tokens of sender
+    */
+    function investorWithdraw(bool keepWrapped)
+        public
+        nonReentrant returns (uint){
```
Comment (Zergity) [inline contracts/TokenVesting.sol:290]: `returns (uint total)`

File: contracts/TokenVesting.sol [hunk 5]
```diff
@@ -266,6 +280,46 @@ contract TokenVesting is Ownable, ReentrancyGuard{
...
+
+        uint256 count = holdersVestingCount[msg.sender];
```
Comment (Zergity) [inline contracts/TokenVesting.sol:292]: unnecessary variable, count only used once

File: contracts/TokenVesting.sol [hunk 6]
```diff
@@ -266,6 +280,46 @@ contract TokenVesting is Ownable, ReentrancyGuard{
...
+        uint256 total = 0;
```
Comment (Zergity) [inline contracts/TokenVesting.sol:293]: remove this declaration after you we already declare in `returns` clause

File: contracts/TokenVesting.sol [hunk 7]
```diff
@@ -266,6 +280,46 @@ contract TokenVesting is Ownable, ReentrancyGuard{
...
+        for (uint i = 0; i < count; i++) {
+            VestingSchedule memory vestingSchedule = getVestingSchedule(computeVestingScheduleIdForAddressAndIndex(msg.sender, i));
```
Comment (Zergity) [inline contracts/TokenVesting.sol:295]: missing this check: `isBeneficiary = msg.sender == vestingSchedule.beneficiary`

**Review by Zergity — CHANGES_REQUESTED** [review]

File: contracts/TokenVesting.sol (hunk 7 above)
Comment (diepmv) [reply contracts/TokenVesting.sol:293]: I've have added this check, thanks

File: contracts/TokenVesting.sol (hunk 5 above)
Comment (diepmv) [reply contracts/TokenVesting.sol]: fixed

File: contracts/TokenVesting.sol (hunk 4 above)
Comment (diepmv) [reply contracts/TokenVesting.sol]: fixed

**Review by Zergity — APPROVED** [review]
please add some unit test for this feature

**Pull request merged**